Add support for symbols in `js_name` and `getter`/`setter`

RunDevelopment opened the pull request:
Fixes #4149.

This PR adds basic support for `js_name = "[Symbol.<name>]"`. It generally works, but still needs to be polished.

TODO list:

- [x] Resolve conflicts. This PR conflicts with #4225, #4202, and #4188. Those PR should be dealt with before this one.
- [x] Fix the bug that `getter = name` doesn't rename imported static getters. See `import_static_symbol_getter` in `crates/cli/tests/reference/symbol.rs`. This might be fixed by #4225.
- [x] Add a symbol name filter. Right now, any name is allowed, but we should only allow known symbol names. This should also blacklist `Symbol.dispose` as [pointed out here](https://github.com/rustwasm/wasm-bindgen/issues/4149#issuecomment-2400851626).
- [x] More testing, especially for the errors users get when using `js_name = Symbol.name` in invalid positions. Users should ideally get an error from the `#[wasm_bindgen]` macro, so users get errors in the IDEs.

Leaving for follow-ons:
- Automatically polyfill missing symbols. Just like #4118, we should polyfill symbols that aren't defined by the JS runtime. We could probably just generalize the `expose_symbol_dispose` method to do this.

RunDevelopment committed:
Add support for symbols in `js_name`

guybedford reviewed (commented):
I like the approach here a lot. Would be happy to review and land further.

In theory the same could apply for private fields or other special cases of compute properties. And if we hit naming conflicts, we could rely on `js_name = "Symbol.iterator"` to disambiguate.

* Do we properly error if attempting to define the same name twice? 
* Can we explicitly ban `Symbol.dispose` being defined on classes, as we define it separately?
* Can we include a test that `js_name = "Symbol.iterator"` can still be used to set the string `MyClass.prototype["Symbol.iterator"]` as another edge case? I believe js_name can be quoted?

guybedford commented:
Rebased this branch onto current main and force-pushed (RunDevelopment/wasm-bindgen@symbols). Latest pushed commit: RunDevelopment/wasm-bindgen@72ae73d6e.

Changes from the original commit:

* Resolved conflicts in `crates/macro-support/src/{ast,parser}.rs` (the old `crates/backend/` is now merged into macro-support), `crates/shared/src/lib.rs`, `crates/cli-support/src/{js,wit}/mod.rs`.
* Integrated with HEAD's additions: `JsNamespace` struct, third `invalid_unused` arg in `attrgen!`, `RegularThis`, `qualified_name`, `private`, `reexport`, the new `FieldLocation`/`FieldInfo` typescript-fields refactor in js/mod.rs, etc.
* Made `BindgenAttrs::js_name()` return `Option<(&JsName, Span)>` and added a small helper macro arm to keep the existing macro structure.
* Reused `property_definition`/`property_accessor`/`property_key` for symbol-aware codegen across class methods, getters, setters, struct fields, and TS field types.
* Properly disambiguate `Symbol.<name>` in optional-chained field access (`?.x` vs `?.[Symbol.x]`).
* Bumped `APPROVED_SCHEMA_FILE_HASH` (no `SCHEMA_VERSION` bump - that's reserved for releases).
* Re-blessed the reference test outputs to match current main's codegen (now generates `.bg.js` for bundler target, etc).
* Added guide coverage for symbol names in `js_name`, `getter`, and `setter` on both imports and Rust exports.
* Added macro UI coverage for invalid symbol positions and invalid computed-key forms.

Status of the prior TODO list:

* Conflicts: resolved and rebased.
* `getter = name` / static getter behavior: existing `import-getter-setter` and `symbol` reference coverage now passes on this branch, including `js_namespace + getter` cases.
* Symbol name filtering: intentionally not added; this keeps accepting any `Symbol.<ident>` rather than maintaining a whitelist.
* `Symbol.dispose`: intentionally not rejected; this keeps the behavior permissive.
* Error coverage: added `crates/macro/ui-tests/symbol-invalid-positions.rs` to pin macro-time diagnostics for invalid positions.

Test status after the latest force-push:

* `cargo fmt --check`
* `cargo test -p wasm-bindgen-macro --test ui`
* `cargo test -p wasm-bindgen-cli --test wasm-bindgen reference`

Generated output for the test case looks like:

```ts
export class Foo {
    private constructor();
    free(): void;
    [Symbol.dispose](): void;
    [Symbol.toPrimitive](): string;
    readonly [Symbol.toStringTag]: string;
}
```

```js
[Symbol.toPrimitive]() { ... }
get [Symbol.toStringTag]() { ... }
// imports:
arg0[Symbol.iterator]()
SomeClass[Symbol.toPrimitive](...)
SomeClass[Symbol.iterator]
```

At this point I believe the review TODO list is handled.

guybedford committed:
Restore dotted js_name semantics, fix clippy/fmt, bless tests

- Fix property_accessor for dotted identifier paths so
  js_name = "prototype.set.call" continues to render as
  obj.prototype.set.call (regression caught by js-sys TypedArray
  coverage tests).
- Derive PartialEq/Eq on ast::Name (required by OperationKind).
- Resolve clippy uninlined_format_args in wit/mod.rs.
- Run cargo fmt.
- Re-bless symbol.* reference test outputs and approve the new
  schema hash.

guybedford committed:
update ui tests

guybedford committed:
Add symbol guide docs and UI tests

guybedford committed:
Update UI test snapshots for CI toolchain

hoodmane commented on crates/cli-support/src/js/mod.rs:
```diff
@@ -5917,27 +5934,79 @@ fn define_export(
+    }
+}
+
+/// Similar the runtime string or symbol value of a property of the given name.
```
```suggestion
/// The runtime string or symbol value of a property of the given name.
```

hoodmane commented on crates/shared/src/lib.rs:
```diff
@@ -242,33 +248,70 @@ pub fn unwrap_function(struct_name: &str) -> String {
+pub enum NameRef<'a> {
+    Identifier(&'a str),
+    Symbol(&'a str),
+}
```
Why do we need `NameRef`? Can't we just use `&Name` everywhere?

guybedford replied:
Refactored to use a pure string convention instead.

guybedford committed:
Collapse Name/AuxName/NameRef machinery to plain strings

Symbol-keyed property names are now plumbed as raw strings end-to-end
in the canonical `"[Symbol.<ident>]"` form. The renderer
(`property_accessor`) recognizes the bracket form and the existing
dotted-path form (`"prototype.set.call"`) directly, removing the
need for the dedicated `Name` enum on the AST and the parallel
`AuxName`/`NameRef` types in the schema and CLI.

Position-specific rejections (structs/enums/types/statics/free
functions/arguments/free imported functions) are now emitted at the
convert sites via the new `BindgenAttrs::js_name_no_symbol` helper
and a small foreign-item-level guard. The parser performs a global
shape check (`validate_computed_key`) so that `"[notSymbol]"` and
`"[Symbol.]"` are rejected at the value's span.

The wasm export-name builders in `wasm-bindgen-shared` now run JS-side
names through a small `export_name_suffix` filter so that bracket-
form `js_name`s map to a stable, identifier-safe wasm symbol on both
sides of the macro/CLI boundary.

This shrinks the diff against `main` substantially (no schema bump,
no new types in shared/macro-support/cli-support) while producing the
same generated JS for symbol-keyed methods, getters, and setters.

hoodmane commented on crates/macro-support/src/ast.rs:
```diff
@@ -572,7 +572,15 @@ impl Export {
-        generated_name.push_str(&self.function.name.to_string());
+        // The JS-side name may contain characters that aren't valid in a
+        // Rust identifier (notably the `[Symbol.<name>]` computed-key form
+        // accepted by `js_name`). Filter to a valid identifier suffix; this
+        // is a no-op for plain identifier names.
```
Doesn't this just delete characters that aren't alphanumeric or underscore? I don't know what is a suffix here.
